src/lib.rs

```diff
@@ -1,411 +1,333 @@
-    let guard_count = (shuffled_alphabet.len() as f32 / GUARD_DIV as f32).ceil() as usize;
+    HashIds::shuffle(&mut alphabet, &salt);
+    let guard_count = (alphabet.len() as f32 / GUARD_DIV).ceil() as usize;
 
-    let t_guards;
+    let guards;
 
-    if shuffled_alphabet.len() < 3 {
-      t_guards = shuffled_separators[..guard_count].to_string();
-      shuffled_separators = shuffled_separators[guard_count..].to_string();
+    if alphabet.len() < 3 {
+      guards = separators[..guard_count].to_vec();
+      separators.drain(..guard_count);
     } else {
-      t_guards = shuffled_alphabet[..guard_count].to_string();
-      shuffled_alphabet = shuffled_alphabet[guard_count..].to_string();
+      guards = alphabet[..guard_count].to_vec();
+      alphabet.drain(..guard_count);
     }
 
     Ok(HashIds {
-      salt: salt,
+      salt: salt.into_boxed_str(),
       min_hash_length: min_length,
-      guards: t_guards,
-      separators: shuffled_separators,
-      alphabet: shuffled_alphabet
+      guards: guards.into_boxed_slice(),
+      separators: separators.into_boxed_slice(),
+      alphabet: alphabet.into_boxed_slice(),
     })
   }
 
-  pub fn new_with_salt_and_min_length(salt: String, min_hash_length: usize) -> Result<HashIds, HashIdsError> {
-    HashIds::new(salt, min_hash_length, DEFAULT_ALPHABET.to_string())
+  pub fn with_min_length(salt: String, min_hash_length: usize) -> Result<HashIds, HashIdsError> {
+    HashIds::with_min_length_and_alphabet(salt, min_hash_length, DEFAULT_ALPHABET)
   }
 
-  pub fn new_with_salt(salt: String) -> Result<HashIds, HashIdsError> {
-    HashIds::new_with_salt_and_min_length(salt, 0)
+  pub fn new(salt: String) -> Result<HashIds, HashIdsError> {
+    HashIds::with_min_length(salt, 0)
   }
 
-
-  fn need_manipulate(slen: usize, alen: usize) -> bool {
-    if slen <= 0 || (((alen/slen) as f32)> SEPARTOR_DIV) {
-      return true;
+  pub fn encode_hex(&self, hex: &str) -> Option<String> {
+    let mut numbers: Vec<u64> = Vec::with_capacity(hex.len() / 12);
+    for chunk in hex.as_bytes().chunks(12) {
+      let mut number = 1;
+      for &ch in chunk {
+        let digit = match ch {
+          b'0'...b'9' => ch - b'0',
+          b'a'...b'f' => ch - b'a' + 10,
+          b'A'...b'F' => ch - b'A' + 10,
+          _ => return None,
+        } as u64;
+        number <<= 4;
+        number |= digit;
+      }
+      numbers.push(number);
     }
-
-    false
+    Some(self.encode(&numbers))
   }
 
-  fn get_min_hash_length(length: usize) -> usize {
-    if length > 0 {
-      return length;
-    } else {
-      return 0;
+  pub fn decode_hex(&self, hash: &str) -> Option<String> {
+    use std::fmt::Write;
+    match self.decode(hash) {
+      Some(numbers) => {
+        let mut result = String::new();
+        let mut buffer = String::new();
+        for number in numbers {
+          write!(buffer, "{:x}", number).unwrap();
+          result.push_str(&buffer[1..])
```

Comment:
The second time through this for loop, you're going to be pushing a buffer that now contains *both* of the numbers you've written to it. That's probably not good, right?